File: JavaRevision/MergeSort.java
import java.util.*;

public class MergeSort {
    public static void print(int arr[]) {
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }

        System.out.println();
    }

    public static int[] merge(int arr[], int low, int mid, int high) {
        List<Integer> list = new ArrayList<>();
        int left = low;
        int right = mid + 1;

        while (left <= mid && right <= high) {
            if (arr[left] <= arr[right]) {
                list.add(arr[left]);
                left++;
            } else {
                list.add(arr[right]);
                right++;
            }
        }

        // if left is remaining
        while (left <= mid) {
            list.add(arr[left]);
            left++;
        }

        // if right is remaining
        while (right <= high) {
            list.add(arr[right]);
            right++;
        }

        // transfer all element from list to arr
        for (int i = low; i <= high; i++) {
            arr[i] = list.get(i - low);
        }

        return arr;
    }

    public static void mergesort(int arr[], int low, int high) {
        if (low >= high) return;
        
        int mid = (low + high) / 2;

        mergesort(arr, low, mid);
        mergesort(arr, mid + 1, high);
        merge(arr, low, mid, high);
    }

    public static void main(String[] args) {
        int arr[] = { 3, 2, 8, 5, 1, 4, 23 };
        print(arr);
        mergesort(arr, 0, arr.length - 1);
        print(arr);
    }
}

File: c/sorting.c/MergeSort.c
#include<stdio.h>
   // merge two subarray
 int merge(int arr[], int l, int mid, int r){

   int n1=mid-l+1;
   int n2=r-mid;
   int left[n1],right[n2];
   int i,j,k;
   for(i=0; i<n1; i++) 
    left[i]=arr[l+i];
    for(j=0; j<n2; j++)
     right[j]=arr[mid+1+j];
    i=0;
    j=0;
    k=l;
    while(i<n1 && j<n2){
        if(left[i]<right[j]){
            arr[k++]=left[i++];
        }
        else 
             arr[k++]=right[j++];
    }
       while(i<n1){
           arr[k++]=left[i++];
       }
         while(i<n1){
            arr[k++]=right[j++]; 
           }
           
 }
 void mergesort(int arr[], int l, int r){
     if(l<r)  {
      int mid=(l+r)/2;
      // sort first & second halves
       mergesort(arr, l, mid);
       mergesort(arr, mid + 1, r);
       merge(arr,l,mid,r);
     } 
}

int main(){
    int arr[]={2,6,3,1,5,7};
    int n = sizeof(arr) / sizeof(int);
    int l=0,r=n-1;
    mergesort(arr,0,n-1);
    for(int i=0;i<n;i++){
      printf("%5d",arr[i]);
    }
    return 0;
}